Sources/Network/NetworkData.swift

```diff
@@ -59,6 +72,32 @@ open class NetworkDataProcedure<Session: URLSessionTaskFactory>: Procedure, Resu
             strongSelf.finish()
         }
 
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToSend), options: [.new, .old], context: &observerContext)
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesSent), options: [.new, .old], context: &observerContext)
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToReceive), options: [.new, .old], context: &observerContext)
+        task?.addObserver(self, forKeyPath: #keyPath(URLSessionTaskProgressProtocol.countOfBytesReceived), options: [.new, .old], context: &observerContext)
+
         task?.resume()
     }
+
+    override open func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
+        guard context == &observerContext else {
+            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
+            return
+        }
+
+        if let keyPath = keyPath, let task = object as? Session.DataTask {
+            switch keyPath {
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToSend):
+                progress.totalUnitCount = task.countOfBytesExpectedToSend
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesSent):
+                progress.completedUnitCount = task.countOfBytesSent
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesExpectedToReceive):
+                progress.totalUnitCount = task.countOfBytesExpectedToReceive
+            case #keyPath(URLSessionTaskProgressProtocol.countOfBytesReceived):
+                progress.completedUnitCount = task.countOfBytesReceived
+            default: break
```

Comment:
Presuming that a data task eventually sets both `countOfBytesExpectedToSend` and `countOfBytesExpectedToReceive`, won't handling both cases as they are above cause progress to increase then regress and then increase again?

i.e.

1. `progress.completedUnitCount` is updated with task.countOfBytesSent until the request is fully sent

2. `progress.totalUnitCount` is **reset** to task.countOfBytesExpectedToReceive

3. `progress.completedUnitCount` is updated with the value of task.countOfBytesReceived, which starts at 0 and increases